build/make/olm.mk

```diff
@@ -0,0 +1,93 @@
+# Copyright (c) 2019-2021 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+#
+
+### generate_olm_bundle: Generate yaml files for building an OLM bundle image
+generate_olm_bundle_yaml: _check_operator_sdk_version _generate_olm_deployment_files
+# Note: operator-sdk provides no way to specify output dir for bundle.Dockerfile so
+# we have to move it manually
+	operator-sdk generate bundle \
+		--deploy-dir deploy/deployment/olm \
+		--output-dir deploy/bundle \
+		--manifests \
+		--metadata && \
+	mv bundle.Dockerfile build/
+
+### generate_olm_bundle_image: build and push DevWorkspace Operator bundle imag
+generate_olm_bundle_image: _print_vars _check_operator_sdk_version
+ifneq ($(INITIATOR),CI)
+ifeq ($(DWO_BUNDLE_IMG),quay.io/devfile/devworkspace-operator-bundle:next)
+	@echo -n "Are you sure you want to push $(DWO_BUNDLE_IMG)? [y/N] " && read ans && [ $${ans:-N} = y ]
+endif
+endif
+	$(DOCKER) build . -t $(DWO_BUNDLE_IMG) -f build/bundle.Dockerfile
+	$(DOCKER) push $(DWO_BUNDLE_IMG)
+
+### generate_olm_index_image: build and push DevWorkspace Operator index image
+generate_olm_index_image: _print_vars _check_skopeo_installed _check_opm_version
+ifneq ($(INITIATOR),CI)
+ifeq ($(DWO_INDEX_IMG),quay.io/devfile/devworkspace-operator-index:next)
+	@echo -n "Are you sure you want to push $(DWO_INDEX_IMG)? [y/N] " && read ans && [ $${ans:-N} = y ]
+endif
+endif
+	export BUNDLE_DIGEST=$$(skopeo inspect docker://$(DWO_BUNDLE_IMG) | jq -r '.Digest') ;\
+	echo "$$BUNDLE_DIGEST" ;\
+	export BUNDLE_IMG=$(DWO_BUNDLE_IMG) ;\
+	export BUNDLE_IMG_DIGEST="$${BUNDLE_IMG%:*}@$${BUNDLE_DIGEST}" ;\
+	opm index add \
+		--bundles "$${BUNDLE_IMG_DIGEST}" \
+		--tag $(DWO_INDEX_IMG) \
+		--container-tool $(DOCKER)
+	$(DOCKER) push $(DWO_INDEX_IMG)
+
+### register_catalogsource: create the catalogsource to make the operator be available on the marketplace
+register_catalogsource: _check_skopeo_installed
+	INDEX_DIGEST=$$(skopeo inspect docker://$(DWO_INDEX_IMG) | jq -r '.Digest')
```

Comment:
I remember Maxim had issues somewhere on QE side that skopeo did not work, just hang for a long time. And he workaround it by using docker inspect.
What if we change it with `docker inspect ${DWO_BUNDLE_IMG} | jq -r '.[0].RepoDigests[0]'`?

Reply:
Discovered a bit more, skopeo works fine for small repos with small tags list, but for repos with millions (number is not precise) it takes a lot of time, e.g. registry-proxy.engineering.redhat.com/rh-osbs/iib:77245
```bash
> skopeo inspect docker://registry-proxy.engineering.redhat.com/rh-osbs/iib:77245 --debug
DEBU[0000] Loading registries configuration "/etc/containers/registries.conf" 
DEBU[0000] Trying to access "registry-proxy.engineering.redhat.com/rh-osbs/iib:77245" 
DEBU[0000] Credentials not found                        
DEBU[0000] Using registries.d directory /etc/containers/registries.d for sigstore configuration 
DEBU[0000]  Using "default-docker" configuration        
DEBU[0000]  No signature storage configuration found for registry-proxy.engineering.redhat.com/rh-osbs/iib:77245 
DEBU[0000] Looking for TLS certificates and private keys in /etc/docker/certs.d/registry-proxy.engineering.redhat.com 
DEBU[0000] GET https://registry-proxy.engineering.redhat.com/v2/ 
DEBU[0000] Ping https://registry-proxy.engineering.redhat.com/v2/ status 401 
DEBU[0000] GET https://registry-proxy.engineering.redhat.com/v2/auth?scope=repository%3Arh-osbs%2Fiib%3Apull 
DEBU[0001] Increasing token expiration to: 60 seconds   
DEBU[0001] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/manifests/77245 
DEBU[0002] Content-Type from manifest GET is "application/vnd.docker.distribution.manifest.list.v2+json" 
DEBU[0002] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/manifests/sha256:8666790758ef80bc7134a5dff81bf2524851774d5a6942bbb6cb4a78363fd3f8 
DEBU[0003] Content-Type from manifest GET is "application/vnd.docker.distribution.manifest.v2+json" 
DEBU[0003] Downloading /v2/rh-osbs/iib/blobs/sha256:3b6d60a018bc515851c948cb951ea90487cca1fd2c9c0cb0bdb24587e734560b 
DEBU[0003] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/blobs/sha256:3b6d60a018bc515851c948cb951ea90487cca1fd2c9c0cb0bdb24587e734560b 
DEBU[0004] Credentials not found                        
DEBU[0004] Using registries.d directory /etc/containers/registries.d for sigstore configuration 
DEBU[0004]  Using "default-docker" configuration        
DEBU[0004]  No signature storage configuration found for registry-proxy.engineering.redhat.com/rh-osbs/iib:77245 
DEBU[0004] Looking for TLS certificates and private keys in /etc/docker/certs.d/registry-proxy.engineering.redhat.com 
DEBU[0004] GET https://registry-proxy.engineering.redhat.com/v2/ 
DEBU[0004] Ping https://registry-proxy.engineering.redhat.com/v2/ status 401 
DEBU[0004] GET https://registry-proxy.engineering.redhat.com/v2/auth?scope=repository%3Arh-osbs%2Fiib%3Apull 
DEBU[0005] Increasing token expiration to: 60 seconds   
DEBU[0005] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/tags/list 
DEBU[0023] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/tags/list?next_page=gAAAAABgrj1yho15L_RZ3jADGQYTiG13JIt55crnbm7MyeC4OfTbUzcg2UBY2InNtvqrzoc9q7U8KbYb_xa-r-tllnO6SwVJNZh5Iw1Jrqurm6TJvw5BX3c%3D&n=50 
DEBU[0024] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/tags/list?next_page=gAAAAABgrj1zT3wgC29PKPVDJhLPi4tabSqSEbSI-LloBfOCFpD3faMWg4dgsL0GENLbs9MExS99R2ntn76FQolTDx--HaaQ-4MODM-u4UKF1O1Dd4ipw9I%3D&n=50 
DEBU[0025] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/tags/list?next_page=gAAAAABgrj1zNpwboA1OZAunudM_E2mYeeJcLzHDO3E66_cLqcffCczj0x_E8xqzvpKGf--OgYAEz6ZutoSEF4ISjWCxwNNTlF7V8Ngp8bybUDxzRnnc4BA%3D&n=50 
DEBU[0027] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/tags/list?next_page=gAAAAABgrj11KDFcYlfSjU1za8igBVhqHeGKA1lzGe6b_bPTbJgdmveN4CccsO-ELx9nK6hqv7TVtFGWrkN_y0zGeTlKCZy9uiGzrtqDNo9OopA9c95ltVI%3D&n=50 
DEBU[0028] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/tags/list?next_page=gAAAAABgrj13Fqd9mkxzAVwedtCzHh8HC2hUz8WUY2pqD-rP-a_9BiwWp3llEN-jWp5jYLzJYaL7AU6Z_N4r1JkK87jl8W5vBIXr9gQI7Vd2RBR4dGOrgdo%3D&n=50 
DEBU[0030] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/tags/list?next_page=gAAAAABgrj14VN08OGMAv9w0HLyirVFQD5txLsczEVAoek8YePLl6xcLjOONDFxKa-IJALHBTfhEcFIesxBkdCulAl_0WvW2UPj_n3ZXx42zcGcNvotliaY%3D&n=50 
DEBU[0031] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/tags/list?next_page=gAAAAABgrj16msZYhXUNQ-SNrlyXxGpnlrp-gA5rzFDpfSyGC5tdB8ulKdUQtkRc4FmElAFhfSgHV6BMHuBrYuewtn-77LFJMCB6dnwA315eJpydZIXj2Lo%3D&n=50 
DEBU[0033] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/tags/list?next_page=gAAAAABgrj173fzBTo1KfnMJb9uLdSdPi4_1af-grcHZn6Yxih-AXzuVdUKN4njgeBByeZmk61s0Gj25Wei0P9rw0cBdHXqB5QhLTqoQpA30FebEgS4mAYQ%3D&n=50 
DEBU[0033] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/tags/list?next_page=gAAAAABgrj17pI6jsAgpp1vMm-zAPvKVsD7tceRcr4Y_CtNdxtzRtrSJJR4uBeMh_CMxiPiCcnT_VdGxPzBAKQX4W-lyfTwud2uqUdQejwyUzmQvYAk7B6s%3D&n=50 
DEBU[0034] GET https://registry-proxy.engineering.redhat.com/v2/rh-osbs/iib/tags/list?next_page=gAA
....
```